pulumi-language-dotnet/main.go

```diff
@@ -559,17 +559,30 @@ func (w *logWriter) LogToUser(val string) (int, error) {
 	return len(val), nil
 }
 
-func (host *dotnetLanguageHost) buildDll() (string, error) {
+// When debugging, we need to build the project, as the debugger does not support running using `dotnet run`.
+// This function will build the project and return the path to the built DLL.
+func (host *dotnetLanguageHost) buildDebuggingDLL(entryPoint string) (string, error) {
 	// If we are running from source, we need to build the project.
 	// Run the `dotnet build` command.  Importantly, report the output of this to the user
 	// (ephemerally) as it is happening so they're aware of what's going on and can see the progress
 	// of things.
 	args := []string{"build", "-nologo", "-o", "bin/pulumi-debugging"}
+	if entryPoint != "" {
+		args = append(args, entryPoint)
+	}
 
 	cmd := exec.Command(host.exec, args...)
-	err := cmd.Run()
+	out, err := cmd.CombinedOutput()
 	if err != nil {
-		return "", errors.Wrapf(err, "failed to build project: %v", err)
+		return "", errors.Wrapf(err, "failed to build project: %v, output: %v", err, string(out))
+	}
+
+	if entryPoint != "" {
+		lastDot := strings.LastIndex(entryPoint, ".")
+		if lastDot == -1 {
+			return "", errors.New("entry point must have a file extension")
```

Comment:
Oh so `req.GetInfo().GetEntryPoint()` is not equivalent to `req.GetProgram()`?  I guess I was wrong in https://github.com/pulumi/pulumi/pull/17402/files#diff-8da99c212de886ea0cb56b605bd48395011b9805cf842e116c8a11dda02af5ffR95 as well then.  So if entrypoint is a folder, I guess we should just look for the `.csproj` under that folder? Or how do I specify that?

I guess https://github.com/pulumi/pulumi-dotnet/blob/main/pulumi-language-dotnet/main.go#L636-L638 is also rather wrong then (though it works fine because `.` is a valid argument there).

Reply:
> Oh so req.GetInfo().GetEntryPoint() is not equivalent to req.GetProgram()?

No they're the same. Refreshed my memory here. EntryPoint will either be a filename _or_ ".". So probably just the if condition that needs to check for != "." rather than not empty.

Reply:
Hrm trying this out folders get put into `program_directory`, not `entry_point`.  It also looks like we generally don't respect folders we get passed in through the engine, so that seems like a bug here, and the default is `.`, so I'll fix that.

Reply:
Yeh so program_directory == the folder made from root + main. entry_point is the file name leftover from main if any, else "."

Reply:
If these docs aren't on the language host protos, they could/should be!

Reply:
Alright, that makes sense to me.  The last two commits should fix this (though I think there's still a bug for the rest of the runtime.  I'll open an issue for that)

Reply:
Actually the current dotnet runtime does seem to handle it correctly (though I'm not entirely sure I understand why), so there's nothing more to do here.

Reply:
> If these docs aren't on the language host protos (and belong there/are true for all languages), they could/should be!

I just checked, and the documentation is actually fine :)

```
    // the absolute path to the directory of the program to execute. Generally, but not required to be,
    // underneath the root directory.
    string program_directory = 2;
    // the entry point of the program, normally '.' meaning to just use the program directory, but can also be
    // a filename inside the program directory. How that filename is interpreted, if at all, is language
    // specific.
    string entry_point = 3;
```